cmake/macros/macro_deal_ii_add_test.cmake

```diff
@@ -412,31 +412,46 @@ function(deal_ii_add_test _category _test_name _comparison_file)
 
       file(MAKE_DIRECTORY ${_test_directory})
 
+      #
+      # Determine whether the test shares a common executable target. This
+      # involves tests with .threads=N. and .mpirun=N. annotation, as well
+      # as tests with parameter files (that might share a common executable
+      # target).
+      #
+      # In this case we have to make sure that concurrently invoking the
+      # test does not accidentally trigger a concurrent build of the
+      # executable target. We ensure this by declaring an additional test
+      # that only builds the shared target / ensures the shared target is
+      # present. All run tests then requires this test target as a "setup
+      # fixture", see
+      # https://cmake.org/cmake/help/latest/prop_test/FIXTURES_REQUIRED.html#prop_test:FIXTURES_REQUIRED
+      #
+      set(_shared_target FALSE)
+      if(NOT "${_n_cpu}${_n_threads}" STREQUAL "00" OR "${_source_file}" MATCHES "(prm|json)$")
+        set(_shared_target TRUE)
+
+        #
+        # Build system-internal target name and final test name for the
+        # "executable" test. We have to make sure that the target and test
+        # names stay the same independent of test name and test category,
+        # thus the rather funny name:
+        #
+        set(_test_executable_target "test_dependency.${_target}.executable")
+        set(_test_executable_full   "test_dependency/${_target}.executable")
+      endif()
+
```

Comment:
Why this complexity? Why don't you just break *any* test into the executable and the execution?

Reply:
@bangerth Because make cannot handle 20k more top level targets. Otherwise I wouldn't add this complexity.

Reply:
Some more context: `make` does not scale in the number of top level targets, it is at least superlinear. (At least not in the way how CMake generates them.) For reference invoking make on any of the top level targets in some of the larger test subdirectories takes about 2-4 seconds. Multiply this by the number of tests and we get a lot of wasted CPU hours on our testers.


In addition we are at the limit of what CDash can handle and process. I think when we hit the 20k test mark we will have to think about another long term solution. So increasing the test artefacts to 30k immediately is an issue...


One possible solution to the make problem would be to move the run and comparison parts entirely into shell scripts. Then, only building the target would be handled in the build system and we could do the above split unconditionally.

The downside of this is that tests would be rerun unconditionally.

Reply:
OK. Would you add a comment that explains the issue with the number of targets in two sentences to the comment at the top of what I marked up?

Reply:
Let me address this in a separate PR.